domain/network/state/movesubnets.go

```diff
@@ -0,0 +1,313 @@
+    WHERE space_uuid IN (SELECT space_uuid FROM from_spaces)
+    UNION ALL
+    -- Application with the space_uuid as default
+    SELECT uuid as application_uuid, space_uuid
+    FROM application
+    WHERE space_uuid IN (SELECT space_uuid FROM from_spaces)
+), space_nodes AS (
+    -- nodes with units belonging to applications with bindings to the specified spaces
+    SELECT DISTINCT u.net_node_uuid AS node_uuid, b.space_uuid
+    FROM unit u
+    JOIN application_bindings b ON u.application_uuid = b.application_uuid
+    UNION
+    -- Machines with positive constraints on the specified spaces
+    SELECT DISTINCT m.net_node_uuid as node_uuid, s.uuid AS space_uuid
+    FROM machine_constraint m
+    JOIN constraint_space cs ON m.constraint_uuid = cs.constraint_uuid
+    JOIN space s ON cs.space = s.name
+    JOIN machine AS m ON m.uuid = m.machine_uuid
+    WHERE s.uuid IN (SELECT space_uuid FROM from_spaces)
+      AND cs.exclude IS FALSE
+), bound_nodes AS (
+    -- Select distinct machines to fetch addresses
+    SELECT DISTINCT node_uuid FROM space_nodes
+), space_addresses AS (
+    -- Get all addresses with their destination space if the subnet is moved
+    SELECT
+        bn.node_uuid,
+        coalesce(ts.space_uuid, s.space_uuid) as space_uuid
+    FROM bound_nodes bn
+    LEFT JOIN ip_address AS a ON bn.node_uuid = a.net_node_uuid
+    LEFT JOIN to_spaces AS ts ON a.subnet_uuid = ts.subnet_uuid
+    LEFT JOIN subnet AS s ON a.subnet_uuid = s.uuid
+), failed_constraints AS (
+    -- Filter out nodes which have an address for every bounded space
+    SELECT *  FROM space_nodes
+	EXCEPT
+	SELECT *  FROM space_addresses
+)
+    SELECT 
+		m.name AS &positiveSpaceConstraintFailure.machine_name,
+		s.name AS &positiveSpaceConstraintFailure.space_name
+FROM failed_constraints AS fc
+JOIN space AS s ON fc.space_uuid = s.uuid
+JOIN machine AS m ON fc.node_uuid = m.net_node_uuid
+`
+	stmt, err := st.Prepare(query, destSpace, movedSubnets, positiveSpaceConstraintFailure{})
+	if err != nil {
+		return nil, errors.Errorf("preparing failed constraint statement: %w,\nquery: %s", err, query)
+	}
+	var failedConstraints []positiveSpaceConstraintFailure
+	err = tx.Query(ctx, stmt, destSpace, movedSubnets).GetAll(&failedConstraints)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("checking constraints: %w", err)
+	}
+
+	return failedConstraints, nil
+}
+
+// checkMovedToMachine verifies negative constraints when subnets are moved to a
+// specific "space" for machines in the system.
+// It returns a slice of negativeSpaceConstraintFailure for failing
+// machines and any error encountered during execution.
+func (st *State) checkMovedToMachine(
+	ctx context.Context,
+	tx *sqlair.TX,
+	movedSubnets uuids,
+	space string,
+) ([]negativeSpaceConstraintFailure, error) {
+
+	type spaceTo struct {
+		Name string `db:"name"`
+	}
+	destSpace := spaceTo{Name: space}
+
+	query := `
+WITH bound_machines AS (    
```

Comment:
Use spaces instead of tabs inside these query strings, GitHub can't render them properly.